document/core/valid/types.rst

```diff
@@ -501,22 +501,22 @@ Limits
 Table Types
 ~~~~~~~~~~~
 
-:math:`\limits~\reftype`
-........................
+:math:`\idxtype~\limits~\reftype`
+.................................
 
-* The limits :math:`\limits` must be :ref:`valid <valid-limits>` within range :math:`2^{32}-1`.
+* The limits :math:`\limits` must be :ref:`valid <valid-limits>` within range :math:`2^{|\idxtype|}`.
 
 * The reference type :math:`\reftype` must be :ref:`valid <valid-reftype>`.
 
 * Then the table type is valid.
 
 .. math::
    \frac{
-     C \vdashlimits \limits : 2^{32} - 1
```

Comment:
I got rid of the `- 1` here. I don't think it's quite right. The limits should be about the _length_ of the sequence, not the _maximum index_. Therefore, it should be valid (semantically) to have an i32 table of length 2^32, or an i64 table of length 2^64.

As far as I know, this shouldn't affect any actual engine because of implementation limits.

Reply:
To _further_ further bolster this point, the spec tests already assert that an i32 table of length 2^32 is valid: https://github.com/mozilla-spidermonkey/memory64/blob/main/test/core/table.wast#L9

Reply:
An i32 table is not allowed to have 2^32 elements, because that value is not representable as an u32, so e.g., table.size wouldn't work. Similarly with i64 tables.

In fact, 2^32 isn't even expressible, since in the binary format, the limit itself is represented using a value of the respective index type. That can't go higher than the corresponding maxuint.

Reply:
Ah, I see. I suppose then that memories only get away with it because they report in terms of pages, not bytes.

Yet another thing that might present a challenge when specifying custom page sizes, I suppose. :slightly_smiling_face: